File: synara/src/app/features/matrix-dto/timeline.ts
/**
 * Timeline item DTOs — product virtualization rows (exhaustive tagged union).
 */

import type { EventId, RoomId, TimelineItemId, UserId } from './ids';
import { parseRelationRef, type RelationRef } from './relation';
import {
  hasForbiddenWireFields,
  isObject,
  optBoolean,
  optString,
  reqNumber,
  reqString,
} from './parseUtil';

export const LOCAL_ECHO_STATES = ['sending', 'sent', 'failed', 'cancelled'] as const;
export type LocalEchoState = typeof LOCAL_ECHO_STATES[number];
const LOCAL_ECHO_SET = new Set<string>(LOCAL_ECHO_STATES);

export function isLocalEchoState(value: unknown): value is LocalEchoState {
  return typeof value === 'string' && LOCAL_ECHO_SET.has(value);
}

export const TIMELINE_ITEM_KINDS = [
  'message',
  'state',
  'membership',
  'reaction_summary',
  'redacted',
  'encrypted_unavailable',
  'date_separator',
  'read_marker',
  'other',
] as const;

export type TimelineItemKind = typeof TIMELINE_ITEM_KINDS[number];

export type TimelineMessageItem = {
  kind: 'message';
  itemId: TimelineItemId;
  eventId: EventId;
  roomId: RoomId;
  sender: UserId;
  originServerTs: number;
  body: string;
  msgtype?: string;
  relatesTo?: RelationRef;
  localEchoState?: LocalEchoState;
  isEdited?: boolean;
  isRedacted?: boolean;
  threadRootId?: EventId;
};

export type TimelineStateItem = {
  kind: 'state';
  itemId: TimelineItemId;
  eventId: EventId;
  roomId: RoomId;
  sender: UserId;
  originServerTs: number;
  stateKey: string;
  stateType: string;
  summary?: string;
};

export type TimelineMembershipItem = {
  kind: 'membership';
  itemId: TimelineItemId;
  eventId: EventId;
  roomId: RoomId;
  sender: UserId;
  originServerTs: number;
  targetUserId: UserId;
  summary: string;
};

export type TimelineReactionSummaryItem = {
  kind: 'reaction_summary';
  itemId: TimelineItemId;
  eventId: EventId;
  roomId: RoomId;
  key: string;
  count: number;
  me?: boolean;
};

export type TimelineRedactedItem = {
  kind: 'redacted';
  itemId: TimelineItemId;
  eventId: EventId;
  roomId: RoomId;
  redactedBy?: EventId;
};

export type TimelineEncryptedUnavailableItem = {
  kind: 'encrypted_unavailable';
  itemId: TimelineItemId;
  eventId: EventId;
  roomId: RoomId;
  reason?: string;
};

export type TimelineDateSeparatorItem = {
  kind: 'date_separator';
  itemId: TimelineItemId;
  dayKey: string;
};

export type TimelineReadMarkerItem = {
  kind: 'read_marker';
  itemId: TimelineItemId;
};

export type TimelineOtherItem = {
  kind: 'other';
  itemId: TimelineItemId;
  eventId?: EventId;
  type?: string;
  summary?: string;
};

export type TimelineItem =
  | TimelineMessageItem
  | TimelineStateItem
  | TimelineMembershipItem
  | TimelineReactionSummaryItem
  | TimelineRedactedItem
  | TimelineEncryptedUnavailableItem
  | TimelineDateSeparatorItem
  | TimelineReadMarkerItem
  | TimelineOtherItem;

export function parseTimelineItem(value: unknown): TimelineItem | null {
  if (!isObject(value) || hasForbiddenWireFields(value)) return null;
  const kind = value.kind;
  if (typeof kind !== 'string') return null;
  const itemId = reqString(value, 'itemId');
  if (itemId === null) return null;

  switch (kind) {
    case 'message': {
      const eventId = reqString(value, 'eventId');
      const roomId = reqString(value, 'roomId');
      const sender = reqString(value, 'sender');
      const originServerTs = reqNumber(value, 'originServerTs');
      const body = reqString(value, 'body');
      const msgtype = optString(value, 'msgtype');
      const isEdited = optBoolean(value, 'isEdited');
      const isRedacted = optBoolean(value, 'isRedacted');
      const threadRootId = optString(value, 'threadRootId');
      if (
        eventId === null ||
        roomId === null ||
        sender === null ||
        originServerTs === null ||
        body === null ||
        msgtype === null ||
        isEdited === null ||
        isRedacted === null ||
        threadRootId === null
      ) {
        return null;
      }
      let relatesTo: RelationRef | undefined;
      if (value.relatesTo !== undefined) {
        const r = parseRelationRef(value.relatesTo);
        if (!r) return null;
        relatesTo = r;
      }
      let localEchoState: LocalEchoState | undefined;
      if (value.localEchoState !== undefined) {
        if (!isLocalEchoState(value.localEchoState)) return null;
        localEchoState = value.localEchoState;
      }
      return {
        kind: 'message',
        itemId,
        eventId,
        roomId,
        sender,
        originServerTs,
        body,
        msgtype,
        relatesTo,
        localEchoState,
        isEdited,
        isRedacted,
        threadRootId,
      };
    }
    case 'state': {
      const eventId = reqString(value, 'eventId');
      const roomId = reqString(value, 'roomId');
      const sender = reqString(value, 'sender');
      const originServerTs = reqNumber(value, 'originServerTs');
      const stateKey = reqString(value, 'stateKey');
      const stateType = reqString(value, 'stateType');
      const summary = optString(value, 'summary');
      if (
        eventId === null ||
        roomId === null ||
        sender === null ||
        originServerTs === null ||
        stateKey === null ||
        stateType === null ||
        summary === null
      ) {
        return null;
      }
      return {
        kind: 'state',
        itemId,
        eventId,
        roomId,
        sender,
        originServerTs,
        stateKey,
        stateType,
        summary,
      };
    }
    case 'membership': {
      const eventId = reqString(value, 'eventId');
      const roomId = reqString(value, 'roomId');
      const sender = reqString(value, 'sender');
      const originServerTs = reqNumber(value, 'originServerTs');
      const targetUserId = reqString(value, 'targetUserId');
      const summary = reqString(value, 'summary');
      if (
        eventId === null ||
        roomId === null ||
        sender === null ||
        originServerTs === null ||
        targetUserId === null ||
        summary === null
      ) {
        return null;
      }
      return {
        kind: 'membership',
        itemId,
        eventId,
        roomId,
        sender,
        originServerTs,
        targetUserId,
        summary,
      };
    }
    case 'reaction_summary': {
      const eventId = reqString(value, 'eventId');
      const roomId = reqString(value, 'roomId');
      const key = reqString(value, 'key');
      const count = reqNumber(value, 'count');
      const me = optBoolean(value, 'me');
      if (eventId === null || roomId === null || key === null || count === null || me === null) {
        return null;
      }
      return { kind: 'reaction_summary', itemId, eventId, roomId, key, count, me };
    }
    case 'redacted': {
      const eventId = reqString(value, 'eventId');
      const roomId = reqString(value, 'roomId');
      const redactedBy = optString(value, 'redactedBy');
      if (eventId === null || roomId === null || redactedBy === null) return null;
      return { kind: 'redacted', itemId, eventId, roomId, redactedBy };
    }
    case 'encrypted_unavailable': {
      const eventId = reqString(value, 'eventId');
      const roomId = reqString(value, 'roomId');
      const reason = optString(value, 'reason');
      if (eventId === null || roomId === null || reason === null) return null;
      return { kind: 'encrypted_unavailable', itemId, eventId, roomId, reason };
    }
    case 'date_separator': {
      const dayKey = reqString(value, 'dayKey');
      if (dayKey === null) return null;
      return { kind: 'date_separator', itemId, dayKey };
    }
    case 'read_marker': {
      return { kind: 'read_marker', itemId };
    }
    case 'other': {
      const eventId = optString(value, 'eventId');
      const type = optString(value, 'type');
      const summary = optString(value, 'summary');
      if (eventId === null || type === null || summary === null) return null;
      return { kind: 'other', itemId, eventId, type, summary };
    }
    default:
      return null;
  }
}

File: crates/synara-core/src/dto/timeline.rs
//! Timeline item DTOs — product virtualization rows (not MatrixEvent graphs).
//!
//! Wire: internally tagged on `kind` (snake_case). Fields camelCase.
//! Body fields are plain-text previews only; no ciphertext dumps required.

use serde::{Deserialize, Serialize};

use super::ids::{EventId, RoomId, TimelineItemId, UserId};
use super::relation::RelationRef;

/// Local-echo / send-queue state for outbound messages.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum LocalEchoState {
    Sending,
    Sent,
    Failed,
    Cancelled,
}

/// Product timeline row (exhaustive tagged union).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum TimelineItem {
    Message(TimelineMessageItem),
    State(TimelineStateItem),
    Membership(TimelineMembershipItem),
    ReactionSummary(TimelineReactionSummaryItem),
    Redacted(TimelineRedactedItem),
    EncryptedUnavailable(TimelineEncryptedUnavailableItem),
    DateSeparator(TimelineDateSeparatorItem),
    ReadMarker(TimelineReadMarkerItem),
    Other(TimelineOtherItem),
}

impl TimelineItem {
    pub fn item_id(&self) -> &str {
        match self {
            Self::Message(i) => &i.item_id,
            Self::State(i) => &i.item_id,
            Self::Membership(i) => &i.item_id,
            Self::ReactionSummary(i) => &i.item_id,
            Self::Redacted(i) => &i.item_id,
            Self::EncryptedUnavailable(i) => &i.item_id,
            Self::DateSeparator(i) => &i.item_id,
            Self::ReadMarker(i) => &i.item_id,
            Self::Other(i) => &i.item_id,
        }
    }

    pub fn kind(&self) -> &'static str {
        match self {
            Self::Message(_) => "message",
            Self::State(_) => "state",
            Self::Membership(_) => "membership",
            Self::ReactionSummary(_) => "reaction_summary",
            Self::Redacted(_) => "redacted",
            Self::EncryptedUnavailable(_) => "encrypted_unavailable",
            Self::DateSeparator(_) => "date_separator",
            Self::ReadMarker(_) => "read_marker",
            Self::Other(_) => "other",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TimelineMessageItem {
    pub item_id: TimelineItemId,
    pub event_id: EventId,
    pub room_id: RoomId,
    pub sender: UserId,
    pub origin_server_ts: u64,
    /// Plain-text preview body only (already redacted/privacy-filtered if needed).
    pub body: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub msgtype: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub relates_to: Option<RelationRef>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub local_echo_state: Option<LocalEchoState>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub is_edited: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub is_redacted: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub thread_root_id: Option<EventId>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TimelineStateItem {
    pub item_id: TimelineItemId,
    pub event_id: EventId,
    pub room_id: RoomId,
    pub sender: UserId,
    pub origin_server_ts: u64,
    pub state_key: String,
    /// Stable type string (e.g. `m.room.name`); not an SDK type object.
    pub state_type: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub summary: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TimelineMembershipItem {
    pub item_id: TimelineItemId,
    pub event_id: EventId,
    pub room_id: RoomId,
    pub sender: UserId,
    pub origin_server_ts: u64,
    pub target_user_id: UserId,
    /// Short membership-change summary for UI (e.g. "joined", "invited").
    pub summary: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TimelineReactionSummaryItem {
    pub item_id: TimelineItemId,
    /// Target event the reactions annotate.
    pub event_id: EventId,
    pub room_id: RoomId,
    pub key: String,
    pub count: u32,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub me: Option<bool>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TimelineRedactedItem {
    pub item_id: TimelineItemId,
    pub event_id: EventId,
    pub room_id: RoomId,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub redacted_by: Option<EventId>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TimelineEncryptedUnavailableItem {
    pub item_id: TimelineItemId,
    pub event_id: EventId,
    pub room_id: RoomId,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TimelineDateSeparatorItem {
    pub item_id: TimelineItemId,
    /// Stable day key for virtualization (e.g. `2026-07-24`).
    pub day_key: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TimelineReadMarkerItem {
    pub item_id: TimelineItemId,
}

/// Safe fallback for unknown/unsupported event shapes — no raw full content dump.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TimelineOtherItem {
    pub item_id: TimelineItemId,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub event_id: Option<EventId>,
    /// Matrix event type string when known (e.g. `m.sticker`).
    #[serde(default, rename = "type", skip_serializing_if = "Option::is_none")]
    pub event_type: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub summary: Option<String>,
}
